Make save state error message more descriptive

### Steps to reproduce

- Go to the project root
- `$ mkdir state.json`
- Run dimooper
- Press `S`

### Observed

```
[ERROR] Is a directory (os error 21)
```

### Expected

A more descriptive error message, because from the currrent one it is not clear what happened.

--- a/src/hardcode.rs
+++ b/src/hardcode.rs
@@ -24,6 +24,9 @@ pub const TTF_FONT_PATH: &'static str = "fonts/minoru.ttf";
 pub const POPUP_FADEOUT_TIME: u32 = 500;
 pub const POPUP_STAY_TIME: u32 = 500;
 
+// TODO(be91501b-fd60-450d-b5c7-5ee4ad261c0c): make STATE_FILE_PATH a Path
+//
+// Consider using https://crates.io/crates/lazy_static
 pub const STATE_FILE_PATH: &'static str = "state.json";
 
 pub const CONFIG_FILE_NAME: &'static str = ".dimooper";

--- a/src/main.rs
+++ b/src/main.rs
@@ -23,6 +23,7 @@ mod ui;
 mod screen;
 mod config;
 mod error;
+mod path;
 
 use midi::PortMidiNoteTracker;
 use ui::Popup;

--- a/src/path.rs
+++ b/src/path.rs
@@ -2,8 +2,8 @@ use std;
 use std::path::{Path, PathBuf};
 
 /// Converts Path into absolute PathBuf. Should be used only for
-/// display in messages since it suppresses any errors and falls back
-/// to returning the relative path.
+/// displaying in messages since it suppresses any errors and falls
+/// back to returning the relative path.
 pub fn display_absolute_path(path: &Path) -> PathBuf {
     std::env::current_dir().map(|d| {
         d.join(path)

--- a/src/path.rs
+++ b/src/path.rs
@@ -1,3 +1,6 @@
+use std;
+use std::path::{Path, PathBuf};
+
 pub fn absolute_path(path: &Path) -> PathBuf {
     std::env::current_dir().map(|d| {
         d.join(path)

--- a/src/path.rs
+++ b/src/path.rs
@@ -1,7 +1,10 @@
 use std;
 use std::path::{Path, PathBuf};
 
-pub fn absolute_path(path: &Path) -> PathBuf {
+/// Converts Path into absolute PathBuf. Should be used only for
+/// display in messages since it suppresses any errors and falls back
+/// to returning the relative path.
+pub fn display_absolute_path(path: &Path) -> PathBuf {
     std::env::current_dir().map(|d| {
         d.join(path)
     }).unwrap_or(path.to_path_buf())

--- a/src/path.rs
+++ b/src/path.rs
@@ -1,0 +1,5 @@
+pub fn absolute_path(path: &Path) -> PathBuf {
+    std::env::current_dir().map(|d| {
+        d.join(path)
+    }).unwrap_or(path.to_path_buf())
+}

--- a/src/screen/looper_screen.rs
+++ b/src/screen/looper_screen.rs
@@ -55,7 +55,9 @@ impl<NoteTracker: MidiNoteTracker> Screen<()> for LooperScreen<NoteTracker> {
                 Event::KeyDown { keycode: Some(Keycode::S), .. } => {
                     match self.looper.save_state_to_file(Path::new(STATE_FILE_PATH)) {
                         Ok(_) => println!("Saved looper state to {}", STATE_FILE_PATH),
-                        Err(e) => println!("[ERROR] {}", e),
+                        Err(e) => println!("[ERROR] Could not save state to {}. Reason: {}",
+                                           STATE_FILE_PATH,
+                                           e),
                     }
                 }
 

--- a/src/screen/looper_screen.rs
+++ b/src/screen/looper_screen.rs
@@ -66,9 +66,12 @@ impl<NoteTracker: MidiNoteTracker> Screen<()> for LooperScreen<NoteTracker> {
                 }
 
                 Event::KeyDown { keycode: Some(Keycode::L), .. } => {
-                    match self.looper.load_state_from_file(Path::new(STATE_FILE_PATH)) {
+                    let state_file_path = Path::new(STATE_FILE_PATH);
+                    match self.looper.load_state_from_file(state_file_path) {
                         Ok(_) => println!("Loaded looper state from {}", STATE_FILE_PATH),
-                        Err(e) => println!("[ERROR] {}", e),
+                        Err(e) => println!("[ERROR] Could not load state from {}. Reason: {}",
+                                           path::absolute_path(state_file_path).display(),
+                                           e),
                     }
                 }
 

--- a/src/screen/looper_screen.rs
+++ b/src/screen/looper_screen.rs
@@ -7,8 +7,7 @@ use screen::Screen;
 use ui::Popup;
 use looper::Looper;
 use hardcode::*;
-use std;
-use std::path::{Path, PathBuf};
+use std::path::Path;
 use traits::*;
 use path;
 
@@ -57,7 +56,7 @@ impl<NoteTracker: MidiNoteTracker> Screen<()> for LooperScreen<NoteTracker> {
 
                 Event::KeyDown { keycode: Some(Keycode::S), .. } => {
                     let state_file_path = Path::new(STATE_FILE_PATH);
-                    let absolute_path = path::absolute_path(state_file_path);
+                    let absolute_path = path::display_absolute_path(state_file_path);
                     match self.looper.save_state_to_file(state_file_path) {
                         Ok(_) => println!("Saved looper state to {}", absolute_path.display()),
                         Err(e) => println!("[ERROR] Could not save state to {}. Reason: {}",
@@ -68,7 +67,7 @@ impl<NoteTracker: MidiNoteTracker> Screen<()> for LooperScreen<NoteTracker> {
 
                 Event::KeyDown { keycode: Some(Keycode::L), .. } => {
                     let state_file_path = Path::new(STATE_FILE_PATH);
-                    let absolute_path = path::absolute_path(state_file_path);
+                    let absolute_path = path::display_absolute_path(state_file_path);
                     match self.looper.load_state_from_file(state_file_path) {
                         Ok(_) => println!("Loaded looper state from {}", absolute_path.display()),
                         Err(e) => println!("[ERROR] Could not load state from {}. Reason: {}",

--- a/src/screen/looper_screen.rs
+++ b/src/screen/looper_screen.rs
@@ -57,20 +57,22 @@ impl<NoteTracker: MidiNoteTracker> Screen<()> for LooperScreen<NoteTracker> {
 
                 Event::KeyDown { keycode: Some(Keycode::S), .. } => {
                     let state_file_path = Path::new(STATE_FILE_PATH);
+                    let absolute_path = path::absolute_path(state_file_path);
                     match self.looper.save_state_to_file(state_file_path) {
-                        Ok(_) => println!("Saved looper state to {}", STATE_FILE_PATH),
+                        Ok(_) => println!("Saved looper state to {}", absolute_path.display()),
                         Err(e) => println!("[ERROR] Could not save state to {}. Reason: {}",
-                                           path::absolute_path(state_file_path).display(),
+                                           absolute_path.display(),
                                            e),
                     }
                 }
 
                 Event::KeyDown { keycode: Some(Keycode::L), .. } => {
                     let state_file_path = Path::new(STATE_FILE_PATH);
+                    let absolute_path = path::absolute_path(state_file_path);
                     match self.looper.load_state_from_file(state_file_path) {
-                        Ok(_) => println!("Loaded looper state from {}", STATE_FILE_PATH),
+                        Ok(_) => println!("Loaded looper state from {}", absolute_path.display()),
                         Err(e) => println!("[ERROR] Could not load state from {}. Reason: {}",
-                                           path::absolute_path(state_file_path).display(),
+                                           absolute_path.display(),
                                            e),
                     }
                 }

--- a/src/screen/looper_screen.rs
+++ b/src/screen/looper_screen.rs
@@ -7,6 +7,7 @@ use screen::Screen;
 use ui::Popup;
 use looper::Looper;
 use hardcode::*;
+use std;
 use std::path::Path;
 use traits::*;
 
@@ -53,10 +54,13 @@ impl<NoteTracker: MidiNoteTracker> Screen<()> for LooperScreen<NoteTracker> {
                 }
 
                 Event::KeyDown { keycode: Some(Keycode::S), .. } => {
-                    match self.looper.save_state_to_file(Path::new(STATE_FILE_PATH)) {
+                    let state_file_path = Path::new(STATE_FILE_PATH);
+                    match self.looper.save_state_to_file(state_file_path) {
                         Ok(_) => println!("Saved looper state to {}", STATE_FILE_PATH),
                         Err(e) => println!("[ERROR] Could not save state to {}. Reason: {}",
-                                           STATE_FILE_PATH,
+                                           std::env::current_dir().map(|d| {
+                                               d.join(state_file_path).display().to_string()
+                                           }).unwrap_or(String::from(STATE_FILE_PATH)),
                                            e),
                     }
                 }

--- a/src/screen/looper_screen.rs
+++ b/src/screen/looper_screen.rs
@@ -8,8 +8,9 @@ use ui::Popup;
 use looper::Looper;
 use hardcode::*;
 use std;
-use std::path::Path;
+use std::path::{Path, PathBuf};
 use traits::*;
+use path;
 
 pub struct LooperScreen<NoteTracker: MidiNoteTracker> {
     looper: Looper<NoteTracker>,
@@ -25,6 +26,7 @@ impl<NoteTracker: MidiNoteTracker> LooperScreen<NoteTracker> {
             quit: false
         }
     }
+
 }
 
 impl<NoteTracker: MidiNoteTracker> Screen<()> for LooperScreen<NoteTracker> {
@@ -58,9 +60,7 @@ impl<NoteTracker: MidiNoteTracker> Screen<()> for LooperScreen<NoteTracker> {
                     match self.looper.save_state_to_file(state_file_path) {
                         Ok(_) => println!("Saved looper state to {}", STATE_FILE_PATH),
                         Err(e) => println!("[ERROR] Could not save state to {}. Reason: {}",
-                                           std::env::current_dir().map(|d| {
-                                               d.join(state_file_path).display().to_string()
-                                           }).unwrap_or(String::from(STATE_FILE_PATH)),
+                                           path::absolute_path(state_file_path).display(),
                                            e),
                     }
                 }

